docs/userGuide/syntax/lists.md

```diff
@@ -109,3 +109,126 @@ first number
 - [x] Item 4
 - ( ) Item 5
 </div>
+
+****Customizing the list appearance:****
+<box type="tip" seamless>
+The list icons can be customized by adding an {icon="..."} at the end of a list item or a list itself.
+</box>
+
+
+<div id="main-example-markbind">
+
+<include src="codeAndOutput.md" boilerplate >
+<variable name="highlightStyle">markdown</variable>
+<variable name="code">
+* item 1 {icon="x"}
+* item 2 {icon="heavy_check_mark"}
+* item 3
+* item 4 {icon="/images/deer.jpg"}
+  
+{icon="fas fa-check"}
+</variable>
+</include>
+
+</div>
+
+<box type="warning">
+
+1. Item-level specifications should only be attached to the list heading or list item, not the list content.
+2. A blank line preceding the content is crucial.
+3. It's important to insert a blank line before list-level specifications.
+4. It is possible to any of the icons supported by MarkBind (e.g., Font Awesome, Octicons, Glyphicons, Images and etc.) as the icon of a list item.
+5. Images will maintain a 1 rem bottom margin.
+</box>
+
+<box type="tip" seamless>
+It is possible to specify icons for nested levels using the icons attribute.
+</box>
+<div id="main-example-markbind">
+
+<include src="codeAndOutput.md" boilerplate >
+<variable name="highlightStyle">markdown</variable>
+<variable name="code">
+* item 1
+* item 2 
+  * item 2.1
+  * item 2.2
+  {icon="x"}
```

Comment:
I forgot to mention this, but a blank line can turn the list into a loose list due to issue #2313. However, this approach allows us to bypass loose list generation when crafting nested lists.

Reply:
So, no blank line needed for list-level specifications for nested lists, but needed for the main list (as per point 3)? It would be better if users don't need to follow such intricate rules.

Reply:
Yes, I recommend using standardized syntax that requires a blank line for all list level configurations. This approach is stable and retains all attributes due to the internal design of markdown-it-attrs, which binds attributes to the nearest block-level element. 

However, due to a design flaw in markdown-it-attrs, an inner list with list level configuration can cause the outer list to become a loose list, as indicated by issue #2313.

For now, I'm using inner list level configurations without blank lines for the demo. While it's not perfect, it works for layer 2 and 3 nested lists. Please be aware this approach may present issues for more complex nested lists due to its lack of consistency.

Reply:
I see. Makes sense from a dev POV, but seems like a random complication from a user POV.  
Is it possible to remove the blank line requirement from _all_ levels, thus making the requirement simpler to users. In addition, we can say that this feature works only up to 3 levels of nesting.

Reply:
Not possible for most outer level:
<img width="358" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/24b45276-ff36-44e0-8a05-97e43285bbbb">
<img width="437" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/299d8868-2f3b-40bb-84a5-0fe0a63d7061">

To address this issue, I can standardize the icon list by enforcing a mandatory blank line. This can be accomplished either by resolving issue #2313 or by eliminating the padding from the 'p' tag for the outermost list level. For the latter approach, the concept of loose or tight lists would no longer apply to icon lists.

Reply:
In that case, let's leave it as it is for now, awaiting others' comments, especially from @ang-zeyu

Reply:
> But otherwise another alternative we can consider is:
> 
> ```
> * item 1 {icon="this will be the outer list icon for the first 3 items"}
> * item 2 a default icon
>   * item 2.1 {icon="this will be all the inner list items' icon"}
>   * item 2.2
> * item 3
> * item 4 {icon="this will be the outer list icon for the last 2 items"}
> * item 5
> ```
> 
> Wdyt?

So, the idea is to keep repeating the 'last known' icon (and settings) until it is changed, and then continue with the changed icon/settings. I like it -- as it is simpler for the user.

Reply:
This is already been implemented 😂

Reply:
> has already already been implemented 😂

@SPWwj When? This is different from the original "item-level overrides list-level" syntax.  In this approach, there is no list-level syntax.

Reply:
This logic is exactly doing the thing you mentioned above 😂
<img width="316" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/50d49a90-8902-4099-b5ba-759a0bfe6ac6">

Reply:
In that case, there should not be any mention of `list-level` in the UG. Example (item 3 below)

```
1. Item-level specifications should be attached only to the list heading or list item, not the list content.
2. Ensuring a blank line precedes the content is crucial.
3. Inserting a blank line before list-level specifications is important.
4. Use any of the icons supported by MarkBind (e.g., Font Awesome, Octicons, Glyphicons, Images, etc.) as the icon of a list item.
5. When using emojis, omit the `::` around the icon name (e.g., use `{icon="emoji"}` instead of `{icon=":emoji:"}`).
```

Reply:
The list-level is quite useful when users have overriding needs.
<img width="523" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/70a0d1af-4eff-497e-83f1-2b0962bf50c9">

The following syntax assumes that the entire list uses the same icon, thereby reducing customizability.

```
* item 1 {icon="this will be the outer list icon for the first 3 items"}
* item 2 a default icon
  * item 2.1 {icon="this will be all the inner list items' icon"}
  * item 2.2
* item 3
* item 4 {icon="this will be the outer list icon for the last 2 items"}
* item 5
```

Reply:
> The following syntax assumes that the entire list uses the same icon, thereby reducing customizability.

I don't think that is the case. Instead of specifying the list-level settings at the end of the list, it takes the setting of the first item as the list-level setting. It can be changed in a subsequent item. Basically, it keeps repeating the 'last used' setting until it is changed, and then keep using the changed setting.  

But I agree that the new syntax is slightly verbose when all items except a few uses a different icon.

Original:
```
* item 1
* item 2
* item 3 {icon="x"}
* item 4

{icon="tick" size="s"}
```

New syntax for the same outcome (note the slight extra verbosity at item 4):
```
* item 1 {icon="tick" size="s"}
* item 2
* item 3 {icon="x"}
* item 4 {icon="tick"}
```

On the positive side, the new syntax is simpler and more intuitive (from a non-programmer POV) e.g., no blank lines needed, and the settings can be understood by reading from the top (without needed to jump to the end to see the default). 

But I think for typical use cases the benefit of the simpler syntax outweighs the cost of extra verbosity.

Reply:
Here's the enhanced version that preserves the overriding feature of the original syntax.

```
* item 1 {icon="tick" size="s"} //Consider if you want to override this instead.
* item 2
* item 3 {icon="x"}
* item 4 {icon="tick" default}
```

Reply:
What's the `default` for? i.e., what's the difference between item 3 and 4?
My thinking was, as the size was never changed in items 2-4, they will all be `size="s"`. But if any item specified a different size, that size will be applied for that item onwards, until a different size is specified in a subsequent item.

Reply:
> Not quite but I think this could work too. 👍

Wdyt @damithc @SPWwj 

> The following syntax assumes that the entire list uses the same icon, thereby reducing customizability.

> What's the default for?

I have the same doubt as @damithc, your current implementation should be plenty fine even without the `default` attribute.

Just make sure to remove:
- `<ul>` related implementations in code
- and list/list-item references in docs

Reply:
> ~~Not quite but I think this could work too. 👍~~

On the other hand, if we have the following it becomes more verbose, but there are probably cases the other way round too...

```
* item 1 {icon="this is the exception"}
* item 2 {icon="this is the default"}
* item 3 {icon="this is the default"}
* item 4 {icon="this is the default"}
* item 5 {icon="this is the default"}
```

vs

```
* item 1 {icon="this is the exception"}
* item 2 {icon="this is the default for every item below"}
* item 3
* item 4
* item 5
```

Reply:
`Default` enhances UX. Note, lists are data, prone to change. When item 2 alters, items 2 and 3 must be updated.
```
* item 1 {icon="this is the exception"} // icon is present
* item 2 {icon="this is the default"}     // icon is present but to change to exception
* item 3 {icon="this is the default"}     // no icon need to set to default, otherwise below will be override by exception what is user forget to set?
* item 4 {icon="this is the default"}     // no icon
* item 5 {icon="this is the default"}    // no icon
```
With "default", there's no specific placement for the icon. Thus, when item 2 changes, only item 2 needs to be updated.
```
* item 1 {icon="this is the exception"} // icon is present
* item 2 {icon="this is the default"}     // no icon, but want to change to exception
* item 3 {icon="this is the default"}     // no icon
* item 4 {icon="this is the default"}     // no icon
* item 5 {icon="this is the default"}    // a default icon
```

Reply:
- Support an `icon-default` attribute. All icons in the list will take up the attributes here (icon, icon-class/width/height/size/etc.)
- `icon-default` is by default the first list item's attributes

Did I get this right? It sounds good to me.

@damithc

Reply:
> Did I get this right? It sounds good to me.

Yes, that sounds good.

Separately, does the `icon-default` of a nested sub-list apply to that sub-list only, or all nested sub-lists at that level?

Reply:
it should apply to that sub-list only otherwise the logic for a list will be very messy:
for example:
```
* list 1 
  * list 1.1 (icon from 5.1) // User will be having a hard time locate for a long complex list
  * list 1.2 (icon from 5.1) //
* list 2 
  * list 2.1
  * list 2.2
* list 3
* list 4
* list 5
  * list 5.1 (default)

```

Reply:
> For the example above, why would the user put the default at the bottom of the list? The more intuitive thing is to specify it as higher in the list as possible (in this case, item 2.1), right?

I don't think place the default at 2.1 or 1.1 or 5.1 matters, the key ideas here is user has to balling where is the default attribute in the list is place at.
To me , it is my personal bias that i believe md should be as straight forward as it is 😂
> That said, I'm fine to scope it to the sub list if the implementation is too complex for the alternative. Particularly so as this is the first version of the feature. It is fine to go with a MVP first.

I understand the concern about users having to specify multiple defaults, which could lead to code duplication. However, much like the previous point, I believe the Markdown (md) should remain as straightforward as possible. The implementation wise do not worry me.

One alternative approach could be to inherit from the parent. So, instead of using the default md dot and white dot, we would populate the list with the default icon, if specified. Each default icon in a nested layer would then take priority over its parent. At the very least, this approach maintains a tree relationship for the default icons.

Reply:
Base on the suggestion above we will encounter something like this:
```
* item 1 {icon="this is the exception"}
  * item 1 {icon="this is the exception"} 
  * item 1 {icon="this is the exception"} 
  * item 1 {icon="this is the exception"}
* item 2 {icon="this is the default"}  
```

then user has to specific:

```
* item 1 {icon="this is the exception"}
  * item 1 {icon="this is the default" default}  
  * item 1 {icon="this is the default"}  
  * item 1 {icon="this is the default"}  
* item 2 {icon="this is the default"}  
```


I think this should be fine.

Reply:
Perhaps the reason why we have this disparity of view is because I view the following _as a list with two levels_, not as _a list with three sub-lists_.  

```
* list 1
  * list 1.1
  * list 1.2
* list 2 
  * list 2.1
  * list 2.2
* list 3
* list 4
* list 5
  * list 5.1
  ```
For the use-case of _changing default icons to ❓  and ✔️_, I would write:
```
* list 1 {icon="question"}
  * list 1.1 (icon="tick"}
  * list 1.2
* list 2 
  * list 2.1
  * list 2.2
* list 3
* list 4
* list 5
  * list 5.1
  ```
output:
```
❓  list 1
  ✔️ list 1.1
  ✔️ list 1.2
❓  list 2 
  ✔️ list 2.1
  ✔️ list 2.2
❓  list 3
❓  list 4
❓  list 5
  ✔️ list 5.1
 ```
Note: There is no need for `default-icon` as the first icon specification at each level is taken as default.

If there is a need to use specific icons for some items (e.g., use ❌ to indicate wrong answer):
```
* list 1 {icon="question"}
  * list 1.1 (icon="x"}
  * list 1.2 {icon="tick" default-icon}
* list 2 
  * list 2.1
  * list 2.2 (icon="x"}
* list 3
* list 4
* list 5
  * list 5.1
```
output:
```
❓  list 1
  ❌ list 1.1
  ✔️ list 1.2
❓  list 2 
  ✔️ list 2.1
  ❌list 2.2
❓  list 3
❓  list 4
❓  list 5
  ✔️ list 5.1
 ```
To me, the above syntax is straight forward to use and understand. But this kind of thing are subjective as it depends on each person's way of viewing things. So, I will not insist on the above. I'm fine if the feature is slightly more verbose for the above cases (i.e., if we go with scoping for each sub level).

Reply:
> output:
> 
> ```
> ❓  list 1
>   ❌ list 1.1
>   ✔️ list 1.2
> ❓  list 2 
>   ✔️ list 2.1
>   ❌list 2.2
> ❓  list 3
> ❓  list 4
> ❓  list 5
>   ✔️ list 5.1
> ```
> 
> To me, the above syntax is straight forward to use and understand. But this kind of thing are subjective as it depends on each person's way of viewing things. So, I will not insist on the above. I'm fine if the feature is slightly more verbose for the above cases (i.e., if we go with scoping for each sub level).

The waterfall model is quite effective, aside from the issue mentioned at: [Comment](https://github.com/MarkBind/markbind/pull/2316#discussion_r1253109949).

There might be a typo here
```
> ```
> ❓  list 1
>   ❌ list 1.1
>   ✔️ list 1.2
> ❓  list 2 
>   ✔️ list 2.1
>   ❌list 2.2
> ❓  list 3
> ❓  list 4
> ❓  list 5
>   ❌ list 5.1
> ```
```

Reply:
> The waterfall model is quite effective, aside from the issue mentioned at: [Comment](https://github.com/MarkBind/markbind/pull/2316#discussion_r1253109949).

I'm not advocating the waterfall model -- this is the same specify-any-item-as-default-and-override-where-needed model you proposed, which I think is a good approach. The user is free to put the `default-icon` in any item, but the top of the list is the natural place to put it (as we read from top to bottom), so I expect most users will do that. 

Secondly, I'm advocating that we treat all level-2 items together rather than separate sub-lists (i.e., only one `default-icon` for all level-2 items). But as I mentioned, I will not insist on it as it is a matter for personal preference.

Reply:
> One alternative approach could be to inherit from the parent. So, instead of using the default md dot and white dot, we would populate the list with the default icon, if specified. Each default icon in a nested layer would then take priority over its parent. At the very least, this approach maintains a tree relationship for the default icons.

Imo its too counterintuitive versus picking one of the below.

> From a user POV, replacing the regular icon of level n with a different icon for the entire list would be a common use case e.g..

> I understand the concern about users having to specify multiple defaults, which could lead to code duplication. However, much like the previous point, I believe the md should remain as straightforward as possible. The implementation wise do not worry me.

Tough one... This would be a really good candidate for a/b testing but we don't have that luxury 😿

Adding a few other considerations (but I'm fine with trying any and pivoting later as well if needed)

1. How often would we expect sublists at the same level to be broken up?
2. How often can we expect extreme cases like:
   ```
   - a
      - a {icon="i-need-to-be-repeated-100-times"}
   - a
      - a {icon="i-need-to-be-repeated-100-times"}
   ... times 100
   ```
   ? (e.g. large student lists, lists of book authors, ...)

   More troublesome but you could also argue for a nunjucks macro in this case..

Reply:
> I actually like the waterfall logic, as it is very intuitive, combined with the defalt attribute can greatly enhance the UX.

I think so too.

> There are circumstances when users want to use the default icon for different nested lists, but at some levels, there may be conflicts. How could we resolve these?

Related to another question I raised earlier. Assuming we allow specifying `default-icon` multiple times, if the user wants to change the default icon at any point, just use another `default-icon` at that point. Would that work?

Reply:
> {icon="i-need-to-be-repeated-100-times"}

If we use nunjucks then all these issue will not be an issue, because it is programable, user do not need to hard code. and everything will be nicely handled by backend.

> How would this work?
> 
> ```
> - a {icon="i-am-the-icon-for-items-1-till-3-because-of-waterfall"}
> - a
> - a
> - a {icon="i-am-the-icon-for-items-4-till-5-because-of-waterfall" default-icon}
> - a
> ```
> Not sure how `default-icon` would have any impact

This is a pure waterfall logic, the default-icon here is redundant.
The conflict lies between the "default" attribute and the idea of treating all "level-2 items" together, as we permit the default attribute to be placed anywhere the user desires.

Both concepts have limitations. Perhaps try the all level waterfall logic, as it uses least attributes?

Reply:
> The conflict lies between the `default` scope list level attribute and the idea of treating all `level-2 items` together, as we permit the `default` attribute to be placed anywhere the user desires.

Yup.

Reply:
Ah... waterfalling the default across sublists (correct me if I'm wrong)

Reply:
> Both concepts have limitations. Perhaps try the all level waterfall logic, as it uses least attributes?

Sounds good to me.

To make sure we're on the same page, what would happen in these cases?

```
- a {icon="icon-1"}
- a
- a
- b {icon="icon-2"}
- b
```


```
- a {icon="icon-1"}
  - a {icon="icon-3"}
  - a
  - a {icon="icon-4"}
- a
- b {icon="icon-2"}
  - a
  - a {icon="icon-5"}
  - a
- b
```


```
- a
  - a
  - a {icon="icon-1"}
- b
  - b
- c {icon="icon-2"}
```

Reply:
My take:
```
- 1 a {icon="icon-1"}
- 1 a
- 1 a
- 2 b {icon="icon-2"}
- 1 b
 ```

```
- 1 a {icon="icon-1"}
  - 3 a {icon="icon-3"}
  - 3 a
  - 4 a {icon="icon-4"}
- 1 a
- 2 b {icon="icon-2"}
  - 3 a
  - 5 a {icon="icon-5"}
  - 3 a
- 1 b
```
```
- a
  - a
  - 1 a {icon="icon-1"}
- b
  - b
- 2 c {icon="icon-2"}
```

Reply:
> > Both concepts have limitations. Perhaps try the all level waterfall logic, as it uses least attributes?
> 
> Sounds good to me.
> 
> To make sure we're on the same page, what would happen in these cases?
> 
> ```
> - a {icon="icon-1"}
> - a 1
> - a 1
> - b {icon="icon-2"}
> - b 2
> ```
> 
> ```
> - a {icon="icon-1"}
>   - a {icon="icon-3"}
>   - a 3
>   - a {icon="icon-4"}
> - a 1
> - b {icon="icon-2"}
>   - a 2
>   - a {icon="icon-5"}
>   - a 5
> - b 2
> ```
> 
> ```
> - a 2
>   - a 5
>   - a {icon="icon-1"}
> - b 2
>   - b 1
> - c {icon="icon-2"}
> ```


> To make sure we're on the same page, what would happen in these cases?

This is a good example 👍. 

Currently, "first and default" is the list's implementation, and we just need to expand this logic to all list levels instead of just at the scope level.

I'm good with both `first and default` and `waterfall` logic.

Should we confirm which one?

Reply:
@SPWwj your answer is very different from mine. 

So my proposal (which is `default + waterfall` ) is as follows:

1. If an item has an icon specified in it, that icon will be used for it. 
1. If the `default-icon` keyword is in the icon specification, that icon will be used as the default for the subsequent items in that level.
1. If an item has no icon specified, the current default icon for that level will be used. If there is no default icon for that level yet, the standard bullet points will be used.
1. The default icon for a level may be changed at any point of the list by using the `default-icon` keyword again. The new default will apply to subsequent items in that level.

I omitted the rule about the icon in the first item becoming the `default-icon` automatically. I think it is cleaner to insist on the `default-icon` keyword, although slightly more work for the user.
I assumed we allow standard bullets to be mixed with custom icons. Not sure if that is feasible. The alternative is 'if one icon is customized, all icons must be customized'. 
As per waterfall strategy, the `default-icon` cannot flow upwards. It applies to items below it only.
The  last rule is already implied by the other rules, but I mentioned it for clarify. 

What do you think?

Reply:
> @SPWwj your answer is very different from mine.
> 
> So my proposal (which is `default + waterfall` ) is as follows:
> 
> 1. If an item has an icon specified in it, that icon will be used for it.
> 2. If the `default-icon` keyword is in the icon specification, that icon will be used as the default for the subsequent items in that level.
> 3. If an item has no icon specified, the current default icon for that level will be used. If there is no default icon for that level yet, the standard bullet points will be used.
> 4. The default icon for a level may be changed at any point of the list by using the `default-icon` keyword again. The new default will apply to subsequent items in that level.


> Not sure if that is feasible. The alternative is 'if one icon is customized, all icons must be customized'. As per waterfall strategy, the `default-icon` cannot flow upwards. It applies to items below it only. The last rule is already implied by the other rules, but I mentioned it for clarify.
> 
> What do you think?

It is doable, but mixing the standard bullests with customs icon may not look appeal because they have different line height from the icon, and we do not want to effect the behavier of original markdown bullet by setting a fix line height for all items.
<img width="92" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/74ee132a-8d04-49e9-9654-31a7c3b208cb">

If we want to apply the waterfall logic to all lists at the same level, here's a suggestion:

 1. If an item has an icon specified in it, that icon will be used for it.
 2. If the `default-icon` keyword is in the icon specification, that icon will be used as the default for the subsequent items in that level.
 3. If an item has no icon specified, the current default icon for that level will be used. If there is no default icon for that level yet,**The nearest icon from the top will be used.** otherwise the **The nearest icon from the bottom will be used.**
 5. The default icon for a level may be changed at any point of the list by using the `default-icon` keyword again. The new default will apply to subsequent items in that level.

The augmented point 3 has actually diminished the importance of the "default" attribute.


Alternatively, we can employ the "first and/or default" approach as demonstrated in your output:
```
- 1 a {icon="icon-1"}
  - 3 a {icon="icon-3"}
  - 3 a
  - 4 a {icon="icon-4"}
- 1 a
- 2 b {icon="icon-2"}
  - 3 a
  - 5 a {icon="icon-5"}
  - 3 a
- 1 b
```

With this approach, users can place "default" at any item within the same level. If not specified, the first icon automatically becomes the default.

Reply:
> the nearest icon from the top will be used, otherwise the The nearest icon from the bottom will be used.

While this works, it is not intuitive/natural. The nearest icon may have been intended as an one-off override. 

On a related note, I feel it would be easier for users to understand (and for us to explain to users) if all defaults are explicitly marked as `default-icon` (and only apply downwards) and all icons without `default-icon` are treated as overrides (and apply to that item only). 

> but mixing the standard bullests with customs icon may not look appeal because they have different line height from the icon, and we do not want to affect the behavier of original markdown bullet by setting a fix line height for all items.

I agree. So, I don't expect users to define lists this way, but we can't leave this case 'undefined' either. If they mix custom icons and standard icons, the list will look ugly, but that's their choice.

Reply:
What if we introduce a default icon as a placeholder for undefined cases? This icon would match the size of the nearest icon in the list, by doing this, we maintain the consistency of undefined cases.

Reply:
> substantial

It look quite well, even with different sizes of icons.

The current version takes the first occurrence as the default. We can use this feature to test and assess the user experience.

<img width="332" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/a2bfcc61-83ed-4e35-9cad-e253d26aaecd">

```
- item 1
- item 2 {size="80px" icon="https://raw.githubusercontent.com/SPWwj/Spirify-Ui/c26edc9581162663d29f4d8f911f98fb7f971ca6/src/assets/svgs/image-error.svg"}
- item 2 { size="80px" icon="https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/event.png?raw=true"}
- item 3 { size="25px" icon="https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/sharing.png?raw=true"}
- item 4 { size="120px" icon="https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/event.png?raw=true"}
- item 5 {icon="https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/attraction.png?raw=true"}
```
While it's feasible to mix standard bullets with customized icons, I wouldn't recommend it. Not only does it potentially compromise aesthetics, but it also introduces complexity due to the icons belonging to different data structures. Any updates made to related components could impact each other, similar to the case when MarkBind removed the 'margin-top' for the 'p' tag, making it challenging to distinguish between loose and tight lists in some scenarios.

Reply:
> It look quite well, even with different sizes of icons.
> 
> The current version takes the first occurrence as the default. We can use this feature to test and assess the user experience.
> 
> <img alt="image" width="332" src="https://user-images.githubusercontent.com/30100720/251745925-a2bfcc61-83ed-4e35-9cad-e253d26aaecd.png">

What's your default placeholder icon for level 1, 2, 3? We should not pick any random images as default placeholders. They should be a close match to standard bullet points. e.g., fontawsome equivalent for black dot, white dot, and black square
* default placeholder for level 1
  * default placeholder for level 2
    * default placeholder for level 3

> While it's feasible to mix standard bullets with customized icons, I wouldn't recommend it.

In that case we can use fontawsome equivalent for standard bullets

Reply:
I hadn't considered that far. As you've noticed, it's merely an error SVG image for an undefined icon. We can certainly determine placeholder images for levels 1, 2, and 3. Since it's an icon list, we have the freedom to decide its appearance, preferably keeping it as close as possible to the standard bullet points.

<img width="402" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/095b17c2-3e86-43bb-ae87-e02e5f760f5b">

Reply:
<img width="203" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/97f04f9c-ca97-4cdd-a35c-697f8784459b">

Reply:
<img width="419" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/7032d22e-a7ba-4a2f-9261-9b4c028fcc02">

We can align the text to the center, ensuring it looks appealing in all situations.

Reply:
This is very unlikely what the user wanted, or would ever want. I'd rather use standard bullets, or if that is not feasible, something that looks like standard bullet points (appearance, size, color etc.) **so that user can clearly see they remain un-customized**. We need not try to auto-customize them because this is not meant to be the final form of them anyway.

Reply:
Is this one better

<img width="184" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/b4fe541e-7d05-4357-a1a5-dc012b8706b8">

Or we go for this one:

![image](https://github.com/MarkBind/markbind/assets/30100720/95ae0f74-af33-43d8-8da1-6430f4858b72)

Reply:
@SPWwj  I prefer the first one (i.e., standard bullets) but you mentioned it has some extra implementation costs, in which case the second one (i.e., similar looking icons from elsewhere) is fine too.

Reply:
> I think users will realize by themselves that it is not a good idea to mix customized icons with standard icons, without us trying to second-guess them.

@damithc @SPWwj how about simply:

1. > but we can't leave this case 'undefined' either 

   Fail hard if we see the user trying to mix the two:
   - waterfall: fail hard if we see the user trying to mix the two in any single level of lists
   - first + default: fail hard if there is no default icon defined.

   - "Fail hard" can be: logging a warning + failing the build
2. If there's really a need for mixing the two, you'd still be able to specify something that looks close to the default from fontawesome/octicon/one of our supported libraries.

Reply:
> default + waterfall


My interpretation of "waterfall" is the [original](https://github.com/MarkBind/markbind/pull/2316#discussion_r1252062287) one with no `default-icon` attribute at all, based on @SPWwj's wording here:

> Both concepts have limitations. Perhaps try the all level waterfall logic, as it uses **least attributes?**

Personally I'd prefer to have no `default-icon` attribute at all if we are going with the waterfall model (otherwise it should be named something like `default-waterfall-icon`).


---

Too recap things too as this thread is getting very long...


- Pure waterfall
   - "Icon shall be X until I see a new one at the same level"
   - design can be more verbose in some cases
      - https://github.com/MarkBind/markbind/pull/2316#discussion_r1253077627 (and comment below that) 
   - for the multiple sublists at the same level issue, it'll be handled by design
- `default-icon`, first `icon` is the default.
   - First `icon` is the default. If its the exception, the author can recontrol it using `default-icon`.
   - for the multiple sublists at the same level issue, we have 2 choices:
      - treat all sublists as one
          - Con: more verbose if sublists should have different defaults
      - treat them separately, defaults should be respecified for every list
	      - I'd prefer this personally between the two options of treating sublists, its simpler to reason with as @SPWwj has mentioned
          - Con: more verbose if all sublists should have the same defaults

Both approaches should be adequate to support simple and common use cases conveniently. For more complex cases, it should not be impossible either, only slightly more verbose.

Reply:
> Too recap things too as this thread is getting very long...

Overall my personal choice would be `first + default + treat sublists separately`.

- Waterfall is convenient but not the most conventional design.
- In the scenario we need to configure different sublists at the same level separately, things are far easier than the other 2 approaches. It is a rarer case, but I don't think we should be making rare cases too difficult to configure either.
   - e.g. sublist A should have icon X but sublist B should have no icon (in this case I realise we'd actually need to support an additional `no-icon` attribute of sorts for the other 2 approaches as well)
   -  e.g. sublist A should have icon X but sublist B should have icon Y. (as easy as waterfall, but harder for `first + default + treat sublists together`)
- It is more troublesome when all sublists share the same icon but not by too much.
   - for small lists (<10), a few ctrl-c + ctrl-v's
   - for larger lists, you'd likely be using nunjucks loops already, and there is nunjucks macros/variables

Reply:
> * convenient

Both demonstrations use FontAwesome😂. We can customize the icons to look as close as possible, or even identical, to standard bullet points.

Reply:
> Overall my personal choice would be first + default + treat sublists separately.
Waterfall is convenient but not the most conventional design.
In the scenario we need to configure different sublists at the same level separately, things are far easier than the other 2 approaches. It is a rarer case, but I don't think we should be making rare cases too difficult to configure either.
e.g. sublist A should have icon X but sublist B should have no icon use the default browser icon (in this case I realise we'd actually need to support an additional no-icon use-browser-default-icon attribute of sorts or icon="use-browser-default" for the other 2 approaches as well)
e.g. sublist A should have icon X but sublist B should have icon Y. (as easy as waterfall, but harder for first + default + treat sublists together)
It is more troublesome when all sublists share the same icon but not by too much.
for small lists (<10), a few ctrl-c + ctrl-v's
for larger lists, you'd likely be using nunjucks loops already, and there is nunjucks macros/variables

Thanks for the recap. To add to this, I believe the choice between 'waterfall' and 'first or default' should be based on the use case.

`Waterfal` would be advantageous when there are many sublists of the same level using the same default icon.
  - Use cases: to-do lists, checklists, progress lists.

`First or default`, on the other hand, is suitable when different sublists need to use different icons.
  - Use cases: team lists, domain-specific lists like subject lists where each subject has different default icons and focus areas.

Lastly, in the later stages if needed, we can certainly incorporate JavaScript into the icons to enhance their functionality. This could potentially replace existing Task lists (from GitHub Flavored Markdown) or Radio-button lists.

Reply:
I think we have different interpretation of waterfall etc. so let's focus on how we are going to explain this to users, as that will be the final delivery format. 

Shall we go with the following then? Not the most efficient in terms of how much extra code user needs to write, but easiest to explain.

1. If an item has an icon specified in it, that icon will be used for it, and subsequent items at that level.
2. If any item in a level is customized, the first item in that level must be customized as well. Otherwise the list will be rejected.

We can replace `level` with `sub-list`, if we want to treat each sub list separately.

Reply:
If our goal is to clarify, then 'first or default' would be the most straightforward and intuitive option. Here's how it works:

- If an item has an icon specified, that icon will be used.
- If an item doesn't have an icon, the default icon will be used.
- The default icon is determined by the 'default' attributes, or by the first item in the list if and only if the 'default' attributes are missing.
- If multiple defaults are present, the first default will be given priority.


> 1. If an item has an icon specified in it, that icon will be used for it, and subsequent items at that level.
> 1. If any item in a level is customized, the first item in that level must be customized as well. Otherwise the list will be rejected.


Regarding the second point, from a user's perspective, this could generate a lot of questions. For instance, what if I have a long list of, say, 10 items, and I only know the icons for items 3, 6, and 9? Does that mean I need to configure items 1, 3, 4, 6, 7, 9, and 10? That doesn't seem very efficient.

Reply:
> Regarding the second point, from a user's perspective, this could generate a lot of questions. For instance, what if I have a long list of, say, 10 items, and I only know the icons for items 3, 6, and 9? Does that mean I need to configure items 1, 3, 4, 6, 7, 9, and 10? That doesn't seem very efficient.

I agree, it's not efficient. A default can make it more efficient. But that is something we can add later. This two-rule version is simpler (to explain, and I assume, to implement) and good enough for the MVP. I don't like it's tendency for more verbose code either, but I'm willing to live with that for now.

The user is going from 'not being able to customize lists at all' to 'able to customize every icon', so it is a big step forward for the user despite the verbosity. I don't think any other Markdown based static site generator has that ability. Users have no cause for complains.

Plus, we want to get out of this analysis-paralysis we are stuck in now, and deliver a working MVP first

Reply:
I'd still prefer `first and default` for the reasons outlined https://github.com/MarkBind/markbind/pull/2316#discussion_r1257008347. Also, no concept of a placeholder icon (instead, failing hard has the benefit of **_forcing_** the author to format the list properly) (instead of having it leak into a deployment).

> we want to get out of this analysis-paralysis we are stuck in now

I strongly agree with this. I think the choice is between the ones here https://github.com/MarkBind/markbind/pull/2316#discussion_r1256988082.

Let's try to avoid mixing the waterfall + first&default models for now.

Reply:
The one I proposed last in https://github.com/MarkBind/markbind/pull/2316#discussion_r1257106600 is same as the first one in https://github.com/MarkBind/markbind/pull/2316#discussion_r1256988082 right?

Reply:
yup

Reply:
> If an item has an icon specified in it, that icon will be used for it, and subsequent items at that level.
> If any item in a level is customized, the first item in that level must be customized as well. Otherwise the list will be rejected.

@ang-zeyu @damithc I've updated the logic to strictly follow a waterfall structure based on the two rules above. Maybe you can try it out and see how it feels. We can adjust and make improvements as necessary.

The checks failed due to an issue with markdown-it-emoji-fixed, but that shouldn't affect other functionalities.

Reply:
@SPWwj I'm getting this from building your docs:

<img src="https://github.com/MarkBind/markbind/assets/3306138/a2dd656d-6d30-4cb1-ad07-5506ec330c8a" width=300 >

Otherwise, while I still prefer `first + default + treat sublists separately`, I'm fine with following waterfall as well, its not a strong preference.

I think we've gathered quite a few number of valuable opinions and ideas by this point too. In the interest of time and preventing decision paralysis, I suggest @damithc make the call at this point.

I'll just make some observations on the differences/additions of your current implementation from the 2 rules since the docs is broken:
- Sublists seem to be treated separately. (I hadn't thought of / noticed this previously, I'll add it to the recap list in https://github.com/MarkBind/markbind/pull/2316#discussion_r1256988082 very shortly for comparison (cc @damithc))
   <img src="https://github.com/MarkBind/markbind/assets/3306138/922a12cb-9575-4d2b-862a-ba4eabc8a9cf" width=300 >

- nearest icon is taken if none is specified
   <img src="https://github.com/MarkBind/markbind/assets/3306138/0dea9b17-69a0-4101-8d1b-32b8d99015e4" width=300 >

Reply:
> The checks failed due to an issue with markdown-it-emoji-fixed, but that shouldn't affect other functionalities.

You should be able to do `import * as emojiData from '../lib/markdown-it/patches/markdown-it-emoji-fixed'` instead.

Reply:
> @SPWwj I'm getting this from building your docs:
> 
> <img alt="" width="300" src="https://user-images.githubusercontent.com/3306138/252020400-a2dd656d-6d30-4cb1-ad07-5506ec330c8a.png">
> Otherwise, while I still prefer `first + default + treat sublists separately`, I'm fine with following waterfall as well, its not a strong preference.
> 
> I think we've gathered quite a few number of valuable opinions and ideas by this point too. In the interest of time and preventing decision paralysis, I suggest @damithc make the call at this point.
> 
> I'll just make some observations on the differences/additions of your current implementation from the 2 rules since the docs is broken:
> 
> * Sublists seem to be treated separately. (I hadn't thought of / noticed this previously, I'll add it to the recap list in [Add Icon and Media Support to List  - Resolves #906 and #2272 #2316 (comment)](https://github.com/MarkBind/markbind/pull/2316#discussion_r1256988082) very shortly for comparison (cc @damithc))
>   <img alt="" width="300" src="https://user-images.githubusercontent.com/3306138/252021601-922a12cb-9575-4d2b-862a-ba4eabc8a9cf.png">
> * nearest icon is taken if none is specified
>   <img alt="" width="300" src="https://user-images.githubusercontent.com/3306138/252021983-0dea9b17-69a0-4101-8d1b-32b8d99015e4.png">

Did you update the branch to the latest [2548b86](https://github.com/MarkBind/markbind/pull/2316/commits/2548b862929e68e3f3af8aaa8b20d7bdbfd2264a)?


As demonstrated by the screenshots in the PR description, it works fine for me. The two rules are also strictly followed.

<img width="553" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/3b4ea552-bde0-4355-805d-fea3df85b420">

I will be back tmr 😂

Reply:
@SPWwj  yes, its the wrong build.

Let me know too if I missed anything:

> `<insert if anything missing>` @SPWwj

Reply:
@ang-zeyu I know why, there is a migration for markdown-it-emoji-fixed.js to ts needed, the npm run clean will remove my markdown-it-emoji-fixed.d.ts file. 

https://github.com/MarkBind/markbind/blob/2548b862929e68e3f3af8aaa8b20d7bdbfd2264a/packages/core/src/lib/markdown-it/patches/markdown-it-emoji-fixed.d.ts

Reply:
It only removes those that have a corresponding `'**/*.js.map'` file (which comes from tsc compiling `.ts`).

You'll need to either `import * as emojiData from '../lib/markdown-it/patches/markdown-it-emoji-fixed'` instead ([ts docs](https://www.typescriptlang.org/docs/handbook/modules.html#import-the-entire-module-into-a-single-variable-and-use-it-to-access-the-module-exports)) or modify the export `module.exports = { content: emojiData }`

Reply:
The d.ts file is an auto generated typing file for ts thus will get removed, this is the reason why the build only works for my local machines😂 The method you mentioned above isn’t what is been used currently? Just that due to tsconfig importing untyped file will throw an error(thus I created d.ts file), I will get back to this soon…

Reply:
> The d.ts file is an auto generated typing file for ts thus will get removed, this is the reason why the build only works for my local machines😂

@SPWwj your `.d.ts` file was manually created so it will be skipped in the `npm run clean` script, that is not the issue. You are attempting to do an es6 named import from a CJS module that exports `module.exports = emojiData` which makes no sense (see above 2 solutions). Typescript definition files do not affect the compilation output here but only type checking.

> The method you mentioned above isn’t what is been used currently?

Please refer to the ts docs ([1](https://www.typescriptlang.org/docs/handbook/modules.html#import-the-entire-module-into-a-single-variable-and-use-it-to-access-the-module-exports), [2](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require)). We have a typescript [migration guide](https://markbind.org/devdocs/devGuide/development/migratingToTypeScript.html#import-export-syntax-reference) in the dg as well. 

> thus I created d.ts file

It's good you've created a definition file for type safety. If you'd like to proceed without it, you'll need to do a regular `const emojiData = require('../lib/markdown-it/patches/markdown-it-emoji-fixed');` so that tsc skips type checking it.

Let me know if its still not clear.

Reply:
https://github.com/MarkBind/markbind/pull/2316#discussion_r1257313936

@damithc based on your review, I assume you are fine with the current approach?

Reply:
> @damithc based on your review, I assume you are fine with the current approach?

Sorry, I missed this one. @SPWwj I wasn't following this thread as it was about the implementation details. What's the 'current approach' referred to here?

Reply:
> Sorry, I missed this one. @SPWwj I wasn't following this thread as it was about the implementation details. What's the 'current approach' referred to here?

Haha, it refers to the first and default model. The waterfall is just another layer on top of it. However, the current implementation is based on the waterfall model, since there hasn't been any objection to it.

Reply:
pinging again @damithc 

@SPWwj If this is ok we can start writing some tests too.

Reply:
> pinging again @damithc

Sorry @ang-zeyu , I'm not sure what's the thing I need to comment on. I'm OK with the current waterfall approach (as per the description in the UG).